### ISSUE TITLE
Generalize MTP sidecar: accept quantized affine heads alongside unquantized BF16

### ISSUE THREAD

**Issue by aosama** [aosama/astronomical#286]
## Summary
Astronomical's MTP sidecar validation and weight binding assume a single format: a one-layer unquantized BF16 head with ~60 tensors. A valid quantized MTP head (e.g. OptiQ INT4 affine, 2339 tensors, U32/BF16 mixed) is rejected as `target_only`. This issue tracks generalized MTP support so Astronomical accepts and executes MTP heads regardless of quantization format.

## Current behavior
- `Qwen3_5MtpSidecarCandidate::validate()` requires `candidate.len() == profiles.len()` with exact dtype/shape match against ~60 BF16 profiles → `Err(())` for any count or dtype deviation.
- `qwen3_5_mtp_tensor_profiles()` generates only unquantized profiles (one `ModelFloat` tensor per projection). A quantized head has 3 tensors per affine projection (weight U32 + scales BF16 + biases BF16).
- `Qwen3_5MtpWeights::bind()` hardcodes `fusion_projection` as `NativeBfloat16` — the one affine projection that doesn't go through `take_quantized_affine_weights`.
- Errors collapse to `()` → only `debug!` logs remain; `mtp_unavailable_reason` provides no human-readable cause.
- Result: quantized MTP heads silently degrade to `target_only`; GUI badge never appears.

## Desired behavior
Astronomical accepts and executes MTP heads in any supported quantization format:
- Unquantized BF16 heads continue working exactly as today.
- Quantized affine heads (U32 packed weights + BF16 scales + BF16 biases, per-module bit widths and group sizes) load with `mtp_runtime_state=active` and produce visible MTP acceleration.
- `mtp_unavailable_reason` provides structured, human-readable diagnostics when validation fails.

### Existing reuse spine
The target model path already has complete affine quantized weight infrastructure:
- `Qwen3_5AffineWeights::Quantized { packed_weight, quantization_scales, quantization_biases, quantization_bits, quantization_group_size }` — consumed by MLX GPU affine kernels.
- `take_quantized_affine_weights()` — resolves `QuantizationProfile` per module, binds weight/scales/biases.
- `Qwen3_5Config.quantization_profile_for_module()` — consults `mtp_quantized_module_profiles` first, then target profiles, then the global default.
- `append_qwen3_5_quantized_affine_tensor_profiles()` — generates U32 weight + BF16 scales/biases profiles for quantized modules.
- `Qwen3_5MtpArtifactCapability` and `Qwen3_5MtpContract` — already separate the contract (depth, semantics) from the weight format.

The MTP path calls `take_quantized_affine_weights` for attention and MoE/MLP projections. Only `fusion_projection` (fc.weight) is hardcoded `NativeBfloat16`. The profile generator calls `append_qwen3_5_mtp_affine_tensor_profiles` which calls `append_qwen3_5_quantized_affine_tensor_profiles`. The config has `mtp_quantized_module_profiles` ready to use.

### Gaps to close

**1. Sidecar format discovery** (`sidecar_declaration.rs`)
Detect whether a sidecar is unquantized BF16 or quantized affine by inspecting tensor dtypes in the safetensors header. A BF16-only header → native MTP format. A mixed U32/BF16 header → quantized affine format. Each format produces different profile shapes and counts.

**2. Quantized MTP profile generation** (`tensor_namespace.rs`, `moe_tensor_spec.rs`, `dense_tensor_spec.rs`)
`qwen3_5_mtp_tensor_profiles()` must generate quantized profiles when the MTP head is quantized. This reuses the existing `append_qwen3_5_quantized_affine_tensor_profiles` (already called from `append_qwen3_5_mtp_affine_tensor_profiles`), but the function needs to receive MTP quantization metadata instead of assuming `ModelFloat` for all MTP tensors.

The profile generator must distinguish:
- Norm tensors (always `ModelFloat` / BF16, never quantized).
- Affine projections (either `ModelFloat` BF16 or `AffineQuantizationFloat` → U32 weight + BF16 scales/biases, depending on per-module quantization profile).
- MoE gate (same choice, depending on profile).

**3. Per-module MTP quantization config plumbing** (`config.rs`, `optiq/config.rs`)
The Ornith config has `mtplx_mtp_quantization = {bits:4, group_size:64}` as a global MTP default, but the `quantization` dict has zero `language_model.mtp.*` per-module entries. Two approaches:
- (a) Per-module overrides in `config.json` `quantization` dict for `language_model.mtp.*` keys (matching target structure). Models with per-module MTP quantization overrides provide them.
- (b) Global default from `mtplx_mtp_quantization` applied to all MTP affine modules when no per-module override exists. Modules stored as BF16 in the sidecar (norms, `fc.weight` when unquantized) are detected via sidecar header inspection.

Approach (b) is needed for backward compatibility — existing models like Ornith don't have per-module MTP overrides. Both can coexist.

**4. Sidecar validation relaxation** (`sidecar_declaration.rs`)
The exact count check `candidate.len() == profiles.len()` must be relaxed. For a quantized head, the sidecar is a superset of the expected canonical names (it contains `mtp.layers.0.mlp.experts.N.xxx.weight/scales/biases` for all 256 experts, which the profile generator must enumerate). Validation should:
- Verify every expected canonical profile has a matching stored tensor.
- Verify stored tensor dtypes match profile expectations (U32 for quantized weights, BF16 for scales/biases/norms).
- Verify stored tensor shapes match profile expectations after accounting for quantization (packed weight shapes differ from unquantized).
- Allow additional stored tensors not covered by profiles (future extensibility).
- Replace `Err(())` with structured error types that surface in `mtp_unavailable_reason`.

**5. Executor binding** (`model.rs`)
`Qwen3_5MtpWeights::bind()` must make `fusion_projection` quantization-aware. Currently:
```rust
let fusion_projection = Qwen3_5AffineWeights::NativeBfloat16 {
    weight: take_tensor(&mut bound_mtp_tensors, "language_model.mtp.fc.weight".to_owned())?,
};
```
Must become:
```rust
let fusion_projection = take_quantized_affine_weights(
    &mut bound_mtp_tensors, qwen3_5_config, "language_model.mtp.fc",
)?;
```
This is a single-line change since `take_quantized_affine_weights` already resolves to `NativeBfloat16` or `Quantized` based on the per-module profile.

**6. MoE expert enumeration for resident MTP** (`moe_tensor_spec.rs`)
For MoE MTP heads, the sidecar enumerates individual experts (`mtp.layers.0.mlp.experts.N.xxx`). Resident MTP validation and weight binding may need those tensors. This does not extend the expert pager or SSD streaming path: there is no MTP sparse-expert layer in the pager, and SSD-paged serving remains target-only.

## Constraint: MTP is not used with SSD streaming

Multi-token prediction is not supported, considered, or executed for SSD streaming. When sparse experts are paged from disk, Astronomical remains target-only even if a valid MTP sidecar is present. The expert pager does not include an MTP sparse-expert layer, and MTP sidecar expert tensors are never planned, read, retained, or executed through the SSD streaming path. Quantized MTP heads in this issue apply only to resident serving.

## Scope

| File | Change |
|---|---|
| `sidecar_declaration.rs` | Format discovery (BF16 vs quantized affine), relaxed validation, structured errors |
| `tensor_namespace.rs` | Quantized MTP profile generation using per-module quantization profiles |
| `moe_tensor_spec.rs` | Resident quantized MoE MTP profiles; MTP experts are not part of SSD streaming |
| `dense_tensor_spec.rs` | Dense MTP profile generation with quantization awareness |
| `config.rs` / `config_accessors.rs` | Populate `mtp_quantized_module_profiles` from `mtplx_mtp_quantization` global default |
| `optiq/config.rs` | `mtp_quantized_module_profiles()` fallback to `mtplx_mtp_quantization` defaults |
| `artifact.rs` | Propagate structured `mtp_unavailable_reason` instead of `debug! + ()` |
| `model.rs` | `fusion_projection` uses `take_quantized_affine_weights` |
| Expert pager / SSD streaming | Out of scope for MTP: no MTP sparse layer in the pager, no SSD-paged MTP execution |

## Acceptance criteria
- [ ] Ornith-1.5-35B-A3B-OptiQ-6bit with `optiq/mtp.safetensors` (2339 tensors, mixed U32/BF16) loads with `mtp_runtime_state=active` when `enabled:true, draft_depth:1`. `GET /v1/status` shows `mtp_effective_execution_draft_depth=1` and the GUI MTP badge appears.
- [ ] Existing unquantized MTP checkpoints still load as before (no regression, `target_only` when no MTP is present).
- [ ] When MTP validation fails, `mtp_unavailable_reason` contains a human-readable cause (e.g. "expected tensor language_model.mtp.fc.weight not found in sidecar", "dtype mismatch: expected BF16 got U32 for tensor mtp.layers.0.self_attn.q_proj.scales") instead of silent fallback.
- [ ] Hermetic tests cover both unquantized (`mtp.safetensors`, BF16-only, ~60 tensors) and quantized (`optiq/mtp.safetensors`, U32+BF16, 2339 tensors) sidecar paths.
- [ ] `fusion_projection` binding respects per-module quantization: unquantized heads bind as `NativeBfloat16`, quantized heads bind as `Quantized`.
- [ ] SSD-paged sparse-expert serving stays target-only even when a valid quantized MTP sidecar is present. The expert pager does not include an MTP sparse layer.

## References
- `crates/model-serving/src/qwen3_5/artifacts/sidecar_declaration.rs` — current exact-match validator
- `crates/model-serving/src/qwen3_5/multi_token_prediction/tensor_namespace.rs` — profile generation (already calls `append_qwen3_5_mtp_affine_tensor_profiles`)
- `crates/model-serving/src/qwen3_5/multi_token_prediction/model.rs` — MTP weight binding (hardcoded `NativeBfloat16` for `fusion_projection`)
- `crates/model-serving/src/qwen3_5/configuration/config_accessors.rs:40` — `quantization_profile_for_module()` already consults `mtp_quantized_module_profiles`
- `crates/model-serving/src/qwen3_5/quantizations/optiq/config.rs:176` — `mtp_quantized_module_profiles()` extracts MTP overrides from config

### Hermetic tests (automated, CI)

The existing test surface already has strong foundations:

- **`mtp_sidecar.rs`** — already tests quantized sidecar fixtures (U32 weight + BF16 scales/biases, `TensorDtype::UInt32` / `TensorDtype::AffineQuantizationFloat`) and asserts exact tensor count, stored-name mapping, and rejection of partial/extra/wrong-dtype/wrong-shape sidecars. Tests both `mtp.safetensors` (root) and `optiq/mtp.safetensors` (nested) paths.

- **`mtp_tensor_namespace.rs`** (dense) — already asserts quantized profile shapes (`U32 [2048, 64]` etc.) for a dense Qwen3.6 config with MTP.

- **`mtp_contract.rs`** — validates the `mtplx_mtp_contract` parsing and field agreement.

New hermetic tests needed for this issue:

1. **Quantized MoE MTP sidecar** — extend `mtp_sidecar.rs` with a test fixture that mirrors the OptiQ quantized MoE format: a small MoE sidecar with multiple expert projections (U32 weight + BF16 scales/biases per projection), shared expert, attention projections, and norms. Assert the validator accepts the full quantized MoE inventory with correct canonical-name mapping, dtype validation, and shape validation after accounting for quantization.

2. **Quantized MoE MTP profile generation** — extend `mtp_tensor_namespace.rs` (or add a `moe/mtp_tensor_namespace.rs`) to assert that `qwen3_5_mtp_tensor_profiles()` generates correct quantized profiles for a MoE config with MTP. Expected: 3 profiles per quantized affine projection (weight U32 + scales BF16 + biases BF16), BF16 for norms, correct expert count, shared expert, shared_expert_gate.

3. **Structured error propagation** — assert that when MTP sidecar validation fails for a quantized head, the `mtp_unavailable_reason` contains a human-readable cause (e.g. "expected tensor language_model.mtp.fc.weight not found in sidecar", "dtype mismatch for language_model.mtp.layers.0.self_attn.q_proj.weight: expected U32 got BF16") instead of the silent `()` fallback.

4. **Config plumbing** — assert that `Qwen3_5Config` populated with `mtplx_mtp_quantization = {bits:4, group_size:64, mode:affine, prequantized:true}` produces correct `mtp_quantized_module_profiles` for MTP affine modules (e.g. `language_model.mtp.fc.weight` resolves to the global MTP quantization default when no per-module override exists).

5. **Binding resolution** — assert that `take_quantized_affine_weights` resolves to `Qwen3_5AffineWeights::Quantized` for MTP modules that have quantization profiles and `NativeBfloat16` for modules without (norms, unquantized `fc.weight`).

### Ignored e2e tests (manual, `#[ignore]`)

The existing `model_artifact_qualification/qwen3_5_moe/mtp/benchmark.rs` already has the exact pattern needed:
- Loads a production MTP model, asserts `MtpRuntimeState::Active`, runs generation with and without MTP, and asserts paired throughput ratio ≥ 1.05 (5% speedup).

New ignored e2e test needed:

6. **Quantized MTP lifecycle and speedup** — `#[ignore]` test that loads Ornith-1.5-35B-A3B-OptiQ-6bit (or equivalent quantized-MTP model) with `acceleration.mtp.enabled:true, draft_depth:1`, asserts `MtpRuntimeState::Active`, runs generation with Romeo and Juliet prompt, and asserts MTP generation throughput ≥ target-only throughput (paired ratio ≥ 1.05). This mirrors the existing `should_keep_representative_mtp_generation_faster_and_target_verified` pattern but exercises the quantized MTP head path.

   The test discovers the model through Astronomical's configured model directories (not hardcoded paths) and requires the model to be locally available. It uses the same fixture source text (`model_metrics_5000_romeo_and_juliet_words.txt`) and follows the 120-second timeout convention for qualification tests.

7. **Quantized MTP weight binding parity** — `#[ignore]` test that loads the quantized MTP model, asserts `MtpRuntimeState::Active`, runs a short generation, and verifies that the MTP weights materialize successfully (no `target_only` fallback) and that greedy output tokens match target-only control (same deterministic sequence when temperature=1 seed is fixed). This proves the quantized head produces identical logits to the unquantized reference.

**Comment by aosama** [comment]
## Design note: format discovery order-of-operations

Currently `artifact.rs` generates MTP profiles **before** opening the sidecar:

```
1. Parse config → generate mtp_tensor_profiles from config
2. Open sidecar candidate
3. Validate candidate against pre-generated profiles
```

With a quantized MTP head this doesn't work because profile generation needs to know **which modules are quantized** (to emit U32 weight + BF16 scales/biases triples vs. a single BF16 weight). The only way to know is to **inspect the sidecar's tensor dtypes first**.

The Ornith model makes this subtle: `fc.weight` and all norms are BF16 (unquantized), but every projection is U32+BF16+BF16 (quantized). We can't apply `mtplx_mtp_quantization = {bits:4}` globally to every MTP module — `fc.weight` must resolve to `NativeBfloat16`.

The proposed approach mirrors what `resolve_unquantized_modules_from_shard_index` already does for the target model:

```
1. Parse config (including mtplx_mtp_quantization as default)
2. Open sidecar candidate
3. Inspect sidecar tensor dtypes → build per-module MTP quantization overrides
   - If module has .weight (U32) + .scales (BF16) + .biases (BF16) → quantized
   - If module has .weight (BF16) only → unquantized
4. Override mtp_quantized_module_profiles from inspection
5. Generate mtp_tensor_profiles from config + overrides
6. Validate candidate against profiles
```

This moves sidecar opening earlier in `artifact.rs` (before profile generation instead of after). The validation step itself doesn't change — we still validate candidate against profiles, just with profiles that match the sidecar's actual format.

**Comment by aosama** [comment]
Recorded in the issue body: multi-token prediction is not supported, considered, or executed for SSD streaming. When sparse experts are paged from disk, serving stays target-only even if a valid MTP sidecar is present. The expert pager does not include an MTP sparse-expert layer.

**Comment by aosama** [comment]
Sidecar validation and resident quantized MTP binding are in #287. MTP remains off unless `acceleration.mtp.enabled` is true. SSD-paged sparse-expert serving stays target-only; the expert pager does not include an MTP layer unless packed `switch_mlp` tensors are present.

A follow-up four-row 4-bit verify kernel is in a stacked PR on that branch. Per-expert `experts.N` paging and 6-bit MoE greedy identity are out of this issue.
